src/Simulation/Core/RuntimeMetadata.cs

```diff
@@ -56,5 +63,69 @@ public class RuntimeMetadata
         /// List of target registers.
         /// </summary>
         public IEnumerable<Qubit> Targets { get; set; } = new List<Qubit>();
+
+        private static bool OnlyOneNull(object? a, object? b) =>
+            (a == null && b != null) || (b == null && a != null);
+        
+        private static bool IsBothNull(object? a, object? b) =>
+            a == null && b == null;
+
+        private static bool ListEquals<T>(IEnumerable<T> a, IEnumerable<T> b) =>
+            IsBothNull(a, b) || (!OnlyOneNull(a, b) && a.SequenceEqual(b));
+
+        public override bool Equals(object? obj)
+        {
+            var other = obj as RuntimeMetadata;
+
+            if (other is null) return false;
+
+            if (!ListEquals<Qubit>(this.Controls, other.Controls)) return false;
+
+            if (!ListEquals<Qubit>(this.Targets, other.Targets)) return false;
+            
+            // If only one children is null, return false
+            if (OnlyOneNull(this.Children, other.Children)) return false;
+            
+            // If both children are not null, compare each child element-wise and return
+            // false if any of them are not equal
+            if (!IsBothNull(this.Children, other.Children))
+            {
+                if (this.Children.Count() != other.Children.Count() ||
+                    this.Children.Zip(other.Children, ListEquals<RuntimeMetadata>).Contains(false))
+                    return false;
+            }
+            
+            return this.Label == other.Label && this.FormattedNonQubitArgs == other.FormattedNonQubitArgs &&
```

Comment:
I would check this first, so you can quickly return `False` if any of these trivial properties fail to be equal. 
Then check the most complex properties one by one.